astropy/wcs/wcs.py

```diff
@@ -2959,8 +2959,18 @@ def slice(self, view, numpy_order=True):
                 else:
                     wcs_new.wcs.crpix[wcs_index] -= iview.start
 
-            nitems = len(builtins.range(self._naxis[wcs_index])[iview])
-            wcs_new._naxis[wcs_index] = nitems
+            try:
+                # range requires integers but the other attributes can also
+                # handle arbitary values, so this needs to be in a try/except.
+                nitems = len(builtins.range(self._naxis[wcs_index])[iview])
+            except TypeError as exc:
+                if 'indices must be integers' not in str(exc):
```

Comment:
Is there any convenience functionality for this style to only catch exceptions of a specified type and with the appropriate message(-part)?